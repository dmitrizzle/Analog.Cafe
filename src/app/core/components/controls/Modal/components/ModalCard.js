import React from "react"

import Card from "../../Card"

export const windowHeight = () =>
  window.innerHeight ||
  document.documentElement.clientHeight ||
  document.body.clientHeight

export default class extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      topOffset: 0,
      topOffsetMax: 20,
      minElementHeight: 80,
      visible: false
    }
  }
  componentWillReceiveProps = () => {
    const element = document.getElementById("modal-card")
    this.setState({ visible: false })
    if (element) {
      window.requestAnimationFrame(() => {
        const elementHeight = element.offsetHeight
        const topOffset = (windowHeight() - elementHeight) / 2
        this.setState({
          topOffset:
            elementHeight >= this.state.minElementHeight
              ? topOffset > this.state.topOffsetMax
                ? topOffset
                : this.state.topOffsetMax
              : this.state.topOffsetMax,
          visible: true
        })
      })

      // const delayReveal = setTimeout(() => {
      //   this.setState({ visible: true })
      //   clearTimeout(delayReveal)
      // }, 150)
    }
  }
  render = () => {
    return (
      <Card
        {...this.props}
        style={{
          opacity: this.state.visible ? 1 : 0,
          transition: "margin 150ms",
          margin: `${this.state.topOffset}px auto 90.1vh`
        }}
        id="modal-card"
      />
    )
  }
}
