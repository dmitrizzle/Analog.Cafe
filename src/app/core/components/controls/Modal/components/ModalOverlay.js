import { connect } from "react-redux"
import React from "react"
import styled from "styled-components"

import { HOST_API, HOST_RUNTIME } from "../../../../../constants"
import { hideModal } from "../../../../store/actions-modal"
import ModalCard from "./ModalCard"

const Overlay = styled.aside`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: ${props => props.theme.layer.card};
  overflow: scroll;
  -webkit-overflow-scrolling: touch;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
`

const ModalOverlay = props => {
  if (!props.modal.hidden && props.modal.status === "ok") {
    import("react-ga").then(ReactGA => {
      ReactGA.modalview(
        props.modal.requested.url
          .replace(HOST_API, "")
          .replace(HOST_RUNTIME, "")
      )
    })
  }
  document.onkeydown = event => {
    if (
      event.keyCode === 27 &&
      !props.modal.info.stubborn &&
      !props.modal.hidden
    )
      props.hideModal()
  }
  return (
    <Overlay
      id="modal-overlay"
      style={{
        display: props.modal.hidden ? "none" : "block"
      }}
      onClick={() => props.hideModal()}
    >
      <ModalCard
        title={props.modal.info.title}
        image={props.modal.info.image}
        text={props.modal.info.text}
        error={props.modal.info.error && props.modal.info.error}
        stubborn={props.modal.info.stubborn}
        headless={props.modal.info.headless}
        buttons={props.modal.info.buttons}
        subscribeForm={props.modal.info.subscribeForm}
        subscribeFormLocation={props.modal.info.subscribeFormLocation}
      />
    </Overlay>
  )
}

const mapStateToProps = state => {
  return {
    modal: state.modal
  }
}
const mapDispatchToProps = dispatch => {
  return {
    hideModal: () => {
      dispatch(hideModal())
    }
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(ModalOverlay)
