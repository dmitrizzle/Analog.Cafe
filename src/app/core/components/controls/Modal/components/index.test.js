import "jest-styled-components"

import { createMockStore } from "redux-test-utils"
import { shallow } from "enzyme"
import { shallowWithStore } from "enzyme-redux"
import React from "react"

import ModalCard from "./ModalCard"
import ModalOverlay, { modalScrollCallback } from "./ModalOverlay"

it("Render ModalOverlay, matches snapshot", () => {
  const store = createMockStore()
  const element = shallowWithStore(<ModalOverlay />, store)
  expect(element).toMatchSnapshot()
})

it("Render ModalCard, matches snapshot", () => {
  const element = shallow(<ModalCard />)
  expect(element).toMatchSnapshot()
})

it("modalScrollCallback function returns null if the target isn't fully scrolled up", () => {
  const callback = () => {}
  expect(
    modalScrollCallback(
      { scrollTop: 1, scrollHeight: 100, clientHeight: 20 },
      callback
    )
  ).toEqual(null)
})

it("modalScrollCallback function returns callback if the target is fully scrolled up", () => {
  const callback = () => "callback"
  expect(
    modalScrollCallback(
      { scrollTop: 80, scrollHeight: 100, clientHeight: 20 },
      callback
    )
  ).toEqual("callback")
})
