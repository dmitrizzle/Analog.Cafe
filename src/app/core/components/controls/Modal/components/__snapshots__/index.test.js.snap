// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ModalCard, matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PureComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "id": "modal-card",
      "style": Object {
        "margin": "0px auto 90.1vh",
        "opacity": 0,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "id": "modal-card",
        "style": Object {
          "margin": "0px auto 90.1vh",
          "opacity": 0,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Render ModalOverlay, matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ModalOverlay) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "hideModal": [Function],
      "modal": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "hideModal": [Function],
        "modal": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
